From b2731a4d91ddecf3cfd03ab17403592cda8e7948 Mon Sep 17 00:00:00 2001
From: Kenan Garibov <kengnatural@gmail.com>
Date: Sat, 6 Feb 2016 17:19:51 +0200
Subject: [PATCH] AndroidRuntime legacy startvm() function returned back to fix
 broken xposed framework. Fix proposed by Stefan Berger
 <s.berger81@gmail.com>.

Change-Id: I5fad896519b2c58086be534a67b69989080712d3
---
 core/jni/AndroidRuntime.cpp              | 26 ++++++++++++++++++++++++++
 include/android_runtime/AndroidRuntime.h |  2 ++
 2 files changed, 28 insertions(+)

diff --git a/core/jni/AndroidRuntime.cpp b/core/jni/AndroidRuntime.cpp
index b8acead..dab7b3b 100644
--- a/core/jni/AndroidRuntime.cpp
+++ b/core/jni/AndroidRuntime.cpp
@@ -433,6 +433,19 @@ void AndroidRuntime::parseExtraOpts(char* extraOptsBuf)
  *
  * Returns 0 on success.
  */
+int AndroidRuntime::startVm(JavaVM** pJavaVM, JNIEnv** pEnv)
+{
+    return startVm(pJavaVM, pEnv, false);
+}
+
+/*
+ * Start the Dalvik Virtual Machine.
+ *
+ * Various arguments, most determined by system properties, are passed in.
+ * The "mOptions" vector is updated.
+ *
+ * Returns 0 on success.
+ */
 int AndroidRuntime::startVm(JavaVM** pJavaVM, JNIEnv** pEnv, bool zygote)
 {
     int result = -1;
@@ -807,6 +820,19 @@ char* AndroidRuntime::toSlashClassName(const char* className)
  * Passes the main function two arguments, the class name and the specified
  * options string.
  */
+void AndroidRuntime::start(const char* className, const char* options)
+{
+    start(className, options, false);
+}
+
+/*
+ * Start the Android runtime.  This involves starting the virtual machine
+ * and calling the "static void main(String[] args)" method in the class
+ * named by "className".
+ *
+ * Passes the main function two arguments, the class name and the specified
+ * options string.
+ */
 void AndroidRuntime::start(const char* className, const char* options, bool zygote)
 {
     ALOGD("\n>>>>>> AndroidRuntime START %s <<<<<<\n",
diff --git a/include/android_runtime/AndroidRuntime.h b/include/android_runtime/AndroidRuntime.h
index e8e869d..ba2f569 100644
--- a/include/android_runtime/AndroidRuntime.h
+++ b/include/android_runtime/AndroidRuntime.h
@@ -64,6 +64,7 @@ public:
 
     int addVmArguments(int argc, const char* const argv[]);
 
+    void start(const char *classname, const char* options);
     void start(const char *classname, const char* options, bool zygote);
 
     void exit(int code);
@@ -117,6 +118,7 @@ private:
     static int startReg(JNIEnv* env);
     void parseExtraOpts(char* extraOptsBuf);
     int startVm(JavaVM** pJavaVM, JNIEnv** pEnv, bool zygote);
+    int startVm(JavaVM** pJavaVM, JNIEnv** pEnv);
 
     Vector<JavaVMOption> mOptions;
     bool mExitWithoutCleanup;
-- 
2.7.0

